### PR TITLE
Functional tests on Windows, closes revault#27

### PR DESCRIPTION
**Pull request by Sam071100**
Addresses Issue #27

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -444,7 +444,7 @@ dependencies = [
 
 [[package]]
 name = "revaultd"
-version = "0.0.3"
+version = "0.0.2"
```
Comment (darosior) [inline Cargo.lock:447]: This shouldn't be modified. You can fix it by just building, i think.

File: tests/fixtures.py [hunk 2]
```diff
@@ -39,9 +39,11 @@ def set_backtrace():
 
 @pytest.fixture(scope="session")
 def test_base_dir():
-    d = os.getenv("TEST_DIR", "/tmp")
+    tempfile.tempdir
```
Comment (darosior) [inline tests/fixtures.py:42]: ??

File: tests/fixtures.py [hunk 3]
```diff
@@ -39,9 +39,11 @@ def set_backtrace():
...
+    a = tempfile.gettempdir()
```
Comment (darosior) [inline tests/fixtures.py:43]: nit: Can you give it a more meaningful name?

File: tests/fixtures.py [hunk 4]
```diff
@@ -39,9 +39,11 @@ def set_backtrace():
...
+    d = os.getenv("TEST_DIR", a)
 
-    directory = tempfile.mkdtemp(prefix="revaultd-tests-", dir=d)
+    directory = tempfile.mkdtemp(prefix="revaultd-", dir=d)
```
Comment (darosior) [inline tests/fixtures.py:46]: I don't think there is a need for removing the `tests-` part anymore

File: tests/test_framework/revaultd.py [hunk 5]
```diff
@@ -35,12 +35,13 @@ def __init__(
         os.makedirs(self.datadir_with_network, exist_ok=True)
 
         bin = os.path.join(
-            os.path.dirname(__file__), "..", "..", "target/debug/revaultd"
+            os.path.dirname(__file__), "..", "..", "target", "debug", "revaultd"
         )
         self.conf_file = os.path.join(datadir, "config.toml")
-        self.cmd_line = [bin, "--conf", f"{self.conf_file}"]
+        self.cmd_line = [bin, "--conf", self.conf_file]
+        # print(self.cmd_line)
```
Comment (darosior) [inline tests/test_framework/revaultd.py:42]: Spurious debug line :)

File: tests/test_framework/revaultd.py [hunk 6]
```diff
@@ -147,7 +148,7 @@ def stop(self, timeout=10):
 
     def cleanup(self):
         try:
-            self.stop()
+            self.stop
```
Comment (darosior) [inline tests/test_framework/revaultd.py:151]: ??

File: tests/test_framework/utils.py [hunk 7]
```diff
@@ -17,18 +18,18 @@
 
 from ephemeral_port_reserve import reserve
 from test_framework import serializations
-from typing import Optional
+from typing import Optional, Tuple
```
Comment (darosior) [inline tests/test_framework/utils.py:21]: I don't think it's necessary?

File: tests/test_framework/utils.py [hunk 8]
```diff
@@ -17,18 +18,18 @@
...
 
 
-TIMEOUT = int(os.getenv("TIMEOUT", 60))
-TEST_DEBUG = os.getenv("TEST_DEBUG", "0") == "1"
+TIMEOUT = int(os.getenv("TIMEOUT", 100))
```
Comment (darosior) [inline tests/test_framework/utils.py:24]: There is no need to change that?

File: tests/test_framework/utils.py [hunk 9]
```diff
@@ -17,18 +18,18 @@
...
+TEST_DEBUG = os.getenv("TEST_DEBUG", "1") == "1"
```
Comment (darosior) [inline tests/test_framework/utils.py:25]: Don't change the default value: if you want debug log traces, set the `TEST_DEBUG` env variable to `1` when running the tests.

File: tests/test_framework/utils.py [hunk 10]
```diff
@@ -17,18 +18,18 @@
...
 EXECUTOR_WORKERS = int(os.getenv("EXECUTOR_WORKERS", 20))
 POSTGRES_USER = os.getenv("POSTGRES_USER", "")
 POSTGRES_PASS = os.getenv("POSTGRES_PASS", "")
 POSTGRES_HOST = os.getenv("POSTGRES_HOST", "localhost")
 POSTGRES_IS_SETUP = POSTGRES_USER and POSTGRES_PASS and POSTGRES_HOST
-VERBOSE = os.getenv("VERBOSE", "0") == "1"
-LOG_LEVEL = os.getenv("LOG_LEVEL", "trace")
+VERBOSE = os.getenv("VERBOSE", "1") == "1"
```
Comment (darosior) [inline tests/test_framework/utils.py:31]: Same

File: tests/test_framework/utils.py [hunk 11]
```diff
@@ -17,18 +18,18 @@
...
+LOG_LEVEL = os.getenv("LOG_LEVEL", "debug")
```
Comment (darosior) [inline tests/test_framework/utils.py:32]: Although i  think it makes sense, this change is unrelated and should *at least* be in another commit

File: tests/test_framework/utils.py [hunk 12]
```diff
@@ -338,36 +340,56 @@ def wrapper(*args, **kwargs):
...
-        this_id = self.next_id
-        resp = self._readobj(sock)
-
-        self.logger.debug("Received response for %s call: %r", method, resp)
-        if "id" in resp and resp["id"] != this_id:
-            raise ValueError(
-                "Malformed response, id is not {}: {}.".format(this_id, resp)
+        if platform.uname()[0] == "Windows":
+            bin = os.path.join(
+            os.path.dirname(__file__), "..", "..", "target", "debug", "revault-cli"
```
Comment (darosior) [inline tests/test_framework/utils.py:345]: Indentation

File: tests/test_framework/utils.py [hunk 13]
```diff
@@ -338,36 +340,56 @@ def wrapper(*args, **kwargs):
...
             )
-        sock.close()
-
-        if not isinstance(resp, dict):
-            raise ValueError(
-                "Malformed response, response is not a dictionary %s." % resp
+            cmd_line = [bin, "--conf", self.config_file, method, *payload]
+            output = subprocess.check_output(cmd_line, shell=True, universal_newlines=True, text=True, encoding='utf-8')
+            self.logger.debug(f"Got this output: {output}")
```
Comment (darosior) [inline tests/test_framework/utils.py:349]: Spurious debug line

File: tests/test_framework/utils.py [hunk 14]
```diff
@@ -338,36 +340,56 @@ def wrapper(*args, **kwargs):
...
+            output = json.loads(output)
+            # output = bytes.decode("utf-8")
```
Comment (darosior) [inline tests/test_framework/utils.py:351]: Spurious debug line

File: tests/test_framework/utils.py [hunk 15]
```diff
@@ -338,36 +340,56 @@ def wrapper(*args, **kwargs):
...
+            self.logger.debug("Received response for %s call: %r", method,output)
```
Comment (darosior) [inline tests/test_framework/utils.py:352]: Please use fstrings in new code.

File: tests/test_framework/utils.py [hunk 16]
```diff
@@ -338,36 +340,56 @@ def wrapper(*args, **kwargs):
...
+
+            if not isinstance(output, dict):
+                raise ValueError(
+                    "Malformed response, response is not a dictionary %s." % output
+                )
+            elif "result" not in output:
+                raise ValueError('Malformed response, "result" missing.')   
+            return output["result"]
```
Comment (darosior) [inline tests/test_framework/utils.py:360]: All this is duplicated with the `else` branch. Can we rather unify it?

File: tests/test_framework/utils.py [hunk 17]
```diff
@@ -460,8 +482,8 @@ def tail(self):
                 self.logs.append(str(line.rstrip()))
                 self.logs_cond.notifyAll()
         self.running = False
-        self.proc.stdout.close()
-        self.proc.stderr.close()
+        # self.proc.stdout.close()
+        # self.proc.stderr.close()
```
Comment (darosior) [inline tests/test_framework/utils.py:486]: Spurious debug lines..

**Review by darosior — COMMENTED** [review]
A first pass, the code is clearly not in a reviewable state. It needs to be cleaned up. Also, please add a short rationale / summary of the approach in the commit message please. I know about the changes here since i've been following with you for the past 2 months but imagine if i were completely blind to most of the reasons for these changes!

Apart from that, great to see progress on Windows support for the functional tests! :tada:

**Comment by darosior** [comment]
Also this commit alone doesn't fix #27 so it should be removed from the title. And the PR fails CI.

File: tests/fixtures.py (hunk 4 above)
Comment (Sam071100) [reply tests/fixtures.py:46]: Yes, It's needed as I tried keeping it but again got the same error of revaultd crashing.

File: tests/test_framework/utils.py (hunk 16 above)
Comment (Sam071100) [reply tests/test_framework/utils.py]: Yep.

File: tests/test_framework/revaultd.py (hunk 6 above)
Comment (darosior) [reply tests/test_framework/revaultd.py:150]: But you *don't call it*! If you don't put parenthesis you are not calling this function. If you have the error
> ''NoneType' object is not callable'

It means `stop` is `None`, which it should never be and you need to fix it.

File: tests/test_framework/utils.py (hunk 11 above)
Comment (darosior) [reply tests/test_framework/utils.py:32]: This wasn't addressed?

File: tests/test_framework/utils.py [hunk 18]
```diff
@@ -338,26 +340,34 @@ def wrapper(*args, **kwargs):
...
-        )
-        sock.sock.send(msg.encode())
-        this_id = self.next_id
-        resp = self._readobj(sock)
-
-        self.logger.debug("Received response for %s call: %r", method, resp)
-        if "id" in resp and resp["id"] != this_id:
-            raise ValueError(
-                "Malformed response, id is not {}: {}.".format(this_id, resp)
+        if platform.uname()[0] == "Windows":
```
Comment (darosior) [inline tests/test_framework/utils.py:343]: nit: you can just use `system()`: https://docs.python.org/3/library/platform.html#platform.system

File: tests/test_framework/utils.py [hunk 19]
```diff
@@ -338,26 +340,34 @@ def wrapper(*args, **kwargs):
...
+            bin = os.path.join(os.path.dirname(__file__), "..", "..", "target", "debug", "revault-cli")
+            cmd_line = [bin, "--conf", self.config_file, method, *payload]
+            resp = subprocess.check_output(cmd_line, shell=True, universal_newlines=True, text=True, encoding='utf-8')
+            resp = json.loads(resp)
+            self.logger.debug(f"Received response for {method} call: {resp}")
```
Comment (darosior) [inline tests/test_framework/utils.py:348]: This is much nicer this way :)

Comment (darosior) [reply tests/test_framework/utils.py:348]: Although the log line could be deduplicated too, but it's a nit

**Review by darosior — COMMENTED** [review]
Much nicer!
As you mentioned privately to me i think it would be helpful to add a section about how to run (part of) the functional tests under Windows in `tests/README.md`

**Comment by darosior** [comment]
Please don't put merge commits in a PR.
